sycl/plugins/level_zero/pi_level0.cpp

```diff
@@ -366,31 +365,103 @@ pi_result _pi_device::initialize() {
   return PI_SUCCESS;
 }
 
-// Crate a new command list to be used in a PI call
 pi_result
-_pi_device::createCommandList(ze_command_list_handle_t *ZeCommandList) {
-  // Create the command list, because in L0 commands are added to
+_pi_queue::resetCommandListFenceEntry(ze_command_list_handle_t ZeCommandList,
+                                      bool MakeAvailable) {
+  // Event has been signaled: If the fence for the associated command list
+  // is signalled, then reset the fence and command list and add them to the
+  // available list for ruse in PI calls.
+  // NOTE: Requires the caller to already have a lock with
+  // ZeCommandListFenceMapMutex.
+  ZE_CALL(zeFenceReset(this->ZeCommandListFenceMap[ZeCommandList]));
+  ZE_CALL(zeCommandListReset(ZeCommandList));
+  if (MakeAvailable) {
+    this->Context->Device->ZeCommandListCacheMutex.lock();
+    this->Context->Device->ZeCommandListCache.push_back(ZeCommandList);
+    this->Context->Device->ZeCommandListCacheMutex.unlock();
+  }
+
+  return PI_SUCCESS;
+}
+
+// Retrieve an available command list to be used in a PI call
+pi_result
+_pi_device::getAvailableCommandList(pi_queue Queue,
+                                    ze_command_list_handle_t *ZeCommandList,
+                                    ze_fence_handle_t *ZeFence) {
+  // Create/Reuse the command list, because in L0 commands are added to
   // the command lists, and later are then added to the command queue.
-  //
-  // TODO: Figure out how to lower the overhead of creating a new list
-  // for each PI command, if that appears to be important.
+  // Each command list is paired with an associated fence to track when the
+  // command list is available for reuse.
+  _pi_result pi_result = PI_OUT_OF_RESOURCES;
   ze_command_list_desc_t ZeCommandListDesc = {};
   ZeCommandListDesc.version = ZE_COMMAND_LIST_DESC_VERSION_CURRENT;
+  ze_fence_desc_t ZeFenceDesc = {};
+
+  // Initally, we need to check if a command list has already been created
+  // on this device that is available for use. If so, then reuse that
+  // L0 Command List and Fence for this PI call.
+  if (Queue->Context->Device->ZeCommandListCache.size() > 0) {
+    Queue->Context->Device->ZeCommandListCacheMutex.lock();
+    *ZeCommandList = Queue->Context->Device->ZeCommandListCache.front();
+    *ZeFence = Queue->ZeCommandListFenceMap[*ZeCommandList];
```

Comment:
Shouldn't `ZeCommandListFenceMapMutex` be locked in order to safely access the `ZeCommandListFenceMap`?

Reply:
Yes, I have a patch in coming that address this part of the code. Thank you for the comment.